src/ibek/ioc_factory.py

```diff
@@ -0,0 +1,225 @@
+
+        return model
+
+    def _process_collections(self, ioc_instance: IOC):
+        """
+        Process all the sub entity collections in the IOC instance
+        """
+        all_entities: List[Entity] = []
+
+        def scan_sub_entities(entity: Entity):
+            # recursive function to scan for SubEntities in an entity
+            definition = entity.__definition__
+
+            # add the parent standard entity
+            all_entities.append(entity)
+
+            # add in SubEntities if any
+            for sub_entity in definition.sub_entities:
+                # find the Entity Class that the SubEntity represents
+                entity_cls = self._entity_classes[sub_entity.type]
+                # get the SubEntity arguments
+                sub_args_dict = sub_entity.model_dump()
+                # jinja render any references to parent Args in the SubEntity Args
+                for key, arg in sub_args_dict.items():
+                    sub_args_dict[key] = UTILS.render(entity, arg)
+                # cast the SubEntity to its concrete Entity subclass
+                cast_entity = entity_cls(**sub_args_dict)
+                # recursively scan the SubEntity for more SubEntities
+                scan_sub_entities(cast_entity)
+
+        for entity in ioc_instance.entities:
+            scan_sub_entities(entity)
+
+        ioc_instance.entities = all_entities
+
+    def _make_entity_model(
+        self, definition: Definition, support: Support
+    ) -> Type[Entity]:
+        """
+        Create an Entity Model from a Definition instance and a Support instance.
+        """
+
+        def add_arg(name, typ, description, default):
+            if default is None:
+                default = PydanticUndefined
+            args[name] = (
+                typ,
+                FieldInfo(description=description, default=default),
+            )
+
+        args: Dict[str, Tuple[type, Any]] = {}
+        validators: Dict[str, Any] = {}
+
+        # fully qualified name of the Entity class including support module
+        full_name = f"{support.module}.{definition.name}"
+
+        # add in each of the arguments as a Field in the Entity
+        for arg in definition.args:
+            full_arg_name = f"{full_name}.{arg.name}"
+            arg_type: Any
+
+            if isinstance(arg, ObjectArg):
+
+                @field_validator(arg.name, mode="after")
+                def lookup_instance(cls, id):
+                    return get_entity_by_id(id)
+
+                validators[full_arg_name] = lookup_instance
+                arg_type = object
+
+            elif isinstance(arg, IdArg):
+                arg_type = str
+
+            elif isinstance(arg, EnumArg):
+                # Pydantic uses the values of the Enum as the options in the schema.
+                # Here we arrange for the keys to be in the schema (what a user supplies)
+                # but the values to be what is rendered when jinja refers to the enum
```

Comment:
In pydantic 2.7.1 I'm getting errors on enums so probably this is all ill advised. I'm not sure I want to spend the time right now to re-discover what the hell I was talking about. Lets assume this will get fixed with [#217](https://github.com/epics-containers/ibek/issues/217)

Woah! this is the first time I have used the GitHub Pull Requests plugin for vscode. It is rather good and I just went looking for that issue number when in fact there is awesome autocompletion making that unnecessary.